File: source/binding/Python/maa/job.py
from .define import *


class Job:
    """异步作业句柄 / Asynchronous job handle

    用于跟踪和管理异步操作的状态，如资源加载、控制器连接等。
    Used to track and manage the status of asynchronous operations
    such as resource loading, controller connection, etc.
    """

    _job_id: MaaId

    def __init__(self, job_id: MaaId, status_func, wait_func):
        self._job_id = job_id
        self._status_func = status_func
        self._wait_func = wait_func

    @property
    def job_id(self) -> int:
        """获取作业 ID / Get job ID

        Returns:
            int: 作业 ID / Job ID
        """
        return int(self._job_id)

    def wait(self) -> "Job":
        """等待作业完成 / Wait for job completion

        阻塞当前线程直到作业完成
        Blocks the current thread until the job is done

        Returns:
            Job: 返回自身，支持链式调用 / Returns self for method chaining
        """
        self._wait_func(self._job_id)
        return self

    @property
    def status(self) -> Status:
        """获取作业状态 / Get job status

        Returns:
            Status: 作业状态 / Job status
        """
        return Status(self._status_func(self._job_id))

    @property
    def done(self) -> bool:
        """判断作业是否已完成 / Check if job is done

        Returns:
            bool: 是否已完成（成功或失败） / Whether done (succeeded or failed)
        """
        return self.status.done

    @property
    def succeeded(self) -> bool:
        """判断作业是否成功 / Check if job succeeded

        Returns:
            bool: 是否成功 / Whether succeeded
        """
        return self.status.succeeded

    @property
    def failed(self) -> bool:
        """判断作业是否失败 / Check if job failed

        Returns:
            bool: 是否失败 / Whether failed
        """
        return self.status.failed

    @property
    def pending(self) -> bool:
        """判断作业是否等待中 / Check if job is pending

        Returns:
            bool: 是否等待中 / Whether pending
        """
        return self.status.pending

    @property
    def running(self) -> bool:
        """判断作业是否运行中 / Check if job is running

        Returns:
            bool: 是否运行中 / Whether running
        """
        return self.status.running


class JobWithResult(Job):
    """带结果的异步作业句柄 / Asynchronous job handle with result

    继承自 Job，额外提供获取作业结果的功能。
    Inherits from Job, additionally provides the ability to get job result.
    """

    def __init__(self, job_id: MaaId, status_func, wait_func, get_func):
        super().__init__(job_id, status_func, wait_func)
        self._get_func = get_func

    def wait(self) -> "JobWithResult":
        """等待作业完成 / Wait for job completion

        Returns:
            JobWithResult: 返回自身，支持链式调用 / Returns self for method chaining
        """
        super().wait()
        return self

    def get(self, wait: bool = False):
        """获取作业结果 / Get job result

        Args:
            wait: 是否在获取结果前等待作业完成，默认为 False。建议先显式调用 wait()（或传入 wait=True），
                确保异步操作已完成后再获取结果 / Whether to wait for job completion before getting result,
                default is False. It's recommended to call wait() first (or pass wait=True) to ensure the
                async operation is finished before getting the result.

        Returns:
            作业执行结果，类型取决于具体作业 / Job execution result, type depends on the specific job
        """

        if wait:
            self.wait()

        return self._get_func(self._job_id)

File: source/binding/Rust/src/job.rs
//! Asynchronous job management for tracking operation status and results.

use crate::{
    common::{MaaId, MaaStatus},
    sys, MaaResult,
};

/// Thread-safe pointer wrapper for FFI closures.
pub struct SendSyncPtr<T>(pub *mut T);
unsafe impl<T> Send for SendSyncPtr<T> {}
unsafe impl<T> Sync for SendSyncPtr<T> {}

impl<T> SendSyncPtr<T> {
    pub fn new(ptr: *mut T) -> Self {
        Self(ptr)
    }
    pub fn get(&self) -> *mut T {
        self.0
    }
}

impl<T> Clone for SendSyncPtr<T> {
    fn clone(&self) -> Self {
        Self(self.0)
    }
}
impl<T> Copy for SendSyncPtr<T> {}

pub type StatusFn = Box<dyn Fn(MaaId) -> MaaStatus + Send + Sync>;
pub type WaitFn = Box<dyn Fn(MaaId) -> MaaStatus + Send + Sync>;

/// An asynchronous operation handle.
///
/// Use this to track the status of controller, resource, and tasker operations.
///
/// # Example
/// ```ignore
/// let job = controller.post_click(100, 200)?;
/// let status = job.wait(); // Blocks until complete
/// if status.succeeded() {
///     println!("Click successful.");
/// }
/// ```
pub struct Job {
    pub id: MaaId,
    status_fn: StatusFn,
    wait_fn: WaitFn,
}

impl Job {
    /// Create a new Job with custom status/wait functions.
    pub fn new(id: MaaId, status_fn: StatusFn, wait_fn: WaitFn) -> Self {
        Self {
            id,
            status_fn,
            wait_fn,
        }
    }

    pub fn for_tasker(tasker_ptr: *mut sys::MaaTasker, id: MaaId) -> Self {
        let ptr = SendSyncPtr::new(tasker_ptr);
        Self {
            id,
            status_fn: Box::new(move |job_id| {
                MaaStatus(unsafe { sys::MaaTaskerStatus(ptr.get(), job_id) })
            }),
            wait_fn: Box::new(move |job_id| {
                MaaStatus(unsafe { sys::MaaTaskerWait(ptr.get(), job_id) })
            }),
        }
    }

    pub fn for_controller(ctrl_ptr: *mut sys::MaaController, id: MaaId) -> Self {
        let ptr = SendSyncPtr::new(ctrl_ptr);
        Self {
            id,
            status_fn: Box::new(move |job_id| {
                MaaStatus(unsafe { sys::MaaControllerStatus(ptr.get(), job_id) })
            }),
            wait_fn: Box::new(move |job_id| {
                MaaStatus(unsafe { sys::MaaControllerWait(ptr.get(), job_id) })
            }),
        }
    }

    pub fn for_resource(res_ptr: *mut sys::MaaResource, id: MaaId) -> Self {
        let ptr = SendSyncPtr::new(res_ptr);
        Self {
            id,
            status_fn: Box::new(move |job_id| {
                MaaStatus(unsafe { sys::MaaResourceStatus(ptr.get(), job_id) })
            }),
            wait_fn: Box::new(move |job_id| {
                MaaStatus(unsafe { sys::MaaResourceWait(ptr.get(), job_id) })
            }),
        }
    }

    /// Block until the operation completes.
    pub fn wait(&self) -> MaaStatus {
        (self.wait_fn)(self.id)
    }

    /// Get the current status without blocking.
    pub fn status(&self) -> MaaStatus {
        (self.status_fn)(self.id)
    }

    pub fn succeeded(&self) -> bool {
        self.status() == MaaStatus::SUCCEEDED
    }

    pub fn failed(&self) -> bool {
        self.status() == MaaStatus::FAILED
    }

    pub fn running(&self) -> bool {
        self.status() == MaaStatus::RUNNING
    }

    pub fn pending(&self) -> bool {
        self.status() == MaaStatus::PENDING
    }

    pub fn done(&self) -> bool {
        let s = self.status();
        s == MaaStatus::SUCCEEDED || s == MaaStatus::FAILED
    }
}

/// An asynchronous operation handle with typed result retrieval.
///
/// Similar to [`Job`] but includes a `get()` method to retrieve the operation result.
pub struct JobWithResult<T> {
    job: Job,
    get_fn: Box<dyn Fn(MaaId) -> MaaResult<Option<T>> + Send + Sync>,
}

impl<T> JobWithResult<T> {
    /// Create a new JobWithResult with custom status/wait/get functions.
    pub fn new(
        id: MaaId,
        status_fn: StatusFn,
        wait_fn: WaitFn,
        get_fn: impl Fn(MaaId) -> MaaResult<Option<T>> + Send + Sync + 'static,
    ) -> Self {
        Self {
            job: Job::new(id, status_fn, wait_fn),
            get_fn: Box::new(get_fn),
        }
    }

    pub fn id(&self) -> MaaId {
        self.job.id
    }

    pub fn wait(&self) -> MaaStatus {
        self.job.wait()
    }

    pub fn status(&self) -> MaaStatus {
        self.job.status()
    }

    pub fn succeeded(&self) -> bool {
        self.job.succeeded()
    }

    pub fn failed(&self) -> bool {
        self.job.failed()
    }

    pub fn running(&self) -> bool {
        self.job.running()
    }

    pub fn pending(&self) -> bool {
        self.job.pending()
    }

    pub fn done(&self) -> bool {
        self.job.done()
    }

    /// Get the operation result.
    ///
    /// # Arguments
    /// * `wait` - If `true`, blocks until the operation completes before getting the result
    pub fn get(&self, wait: bool) -> MaaResult<Option<T>> {
        if wait {
            self.wait();
        }
        (self.get_fn)(self.job.id)
    }
}

// === Type Aliases for Specialized Jobs ===

/// Controller operation job.
///
/// Returned by controller methods like `post_click()`, `post_swipe()`.
pub type CtrlJob = Job;

/// Resource loading job.
///
/// Returned by resource methods like `post_bundle()`.
pub type ResJob = Job;

/// Task job with result retrieval.
///
/// Returned by `Tasker::post_task()`.
pub type TaskJobWithResult = JobWithResult<crate::common::TaskDetail>;

/// Recognition job with result retrieval.
///
/// Returned by `Tasker::post_recognition()`.
pub type RecoJobWithResult = JobWithResult<crate::common::RecognitionDetail>;

/// Action job with result retrieval.
///
/// Returned by `Tasker::post_action()`.
pub type ActionJobWithResult = JobWithResult<crate::common::ActionDetail>;

pub fn tasker_ptr(ptr: *mut sys::MaaTasker) -> SendSyncPtr<sys::MaaTasker> {
    SendSyncPtr::new(ptr)
}
